clippy_lints/src/utils/internal_lints.rs

```diff
@@ -345,3 +368,13 @@ fn is_trigger_fn(fn_kind: FnKind<'_>) -> bool {
         FnKind::Closure(..) => false,
     }
 }
+
+declare_lint_pass!(DefaultLint => [DEFAULT_LINT]);
+
+impl EarlyLintPass for DefaultLint {
+    fn check_mac(&mut self, cx: &EarlyContext<'_>, mac: &ast::Mac) {
+        if mac.path == sym!(declare_lint) {
+            println!("{:?}", mac);
+        }
+    }
```

Comment:
I think it is easier to check the expanded code, instead of the macro. The lint declaration macros expand to this:

```rust
pub static TEST_LINT: &::rustc_session::lint::Lint =
    &::rustc_session::lint::Lint{name: &"clippy::TEST_LINT",
                                 default_level: ::rustc_session::lint::Warn,
                                 desc: "",
                                 edition_lint_opts: None,
                                 report_in_external_macro: true,
                                 future_incompatible: None,
                                 is_plugin: true,};
```

where you then can check the `desc` field.

Matching the `public static ... = &Lint{..}` is already done here

https://github.com/rust-lang/rust-clippy/blob/05cb0df748dbc14c43dea4d627d68721a52861c6/clippy_lints/src/utils/internal_lints.rs#L170-L173

you just have to add the check for the `desc` field.

Reply:
oh nice thank you! This is probably a rookie mistake but I'm having some trouble accessing the actual `Lint` struct. I can see it within the output of `dbg!(ty)` but I can't figure out how to get to the `desc` field.

Reply:
The constructor of the `Lint` struct is not in the `ty`, but in the third argument, which is a `BodyId`.

You can look at it like this:

```rust
   pub static [mut] NAME: T = T { ... }; 
// ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ hir::Item
//     ^^^^^^ |   |       |   |       |  hir::ItemKind::Static(ty, mutability, body_id)
//            ^^^^^       |   |       |  mutability
//                        ^   |       |  ty
//                            ^^^^^^^^^  body (accessible through body_id)
```

The HIR is build with IDs referring the the underlying `Expr`s/`Item`s/`Body`s/.... So you have to get the body with [`hir::map::body`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/map/struct.Map.html#method.body). You can get the `hir::map` through `cx.tcx.hir()`. The `Expr` you then have to check is the [`value`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Body.html#structfield.value) field of the `hir::Body` struct.

Reply:
thanks, that helped a ton!